src/System.Net.HttpListener/src/System/Net/Windows/HttpListener.Windows.cs

```diff
@@ -1341,6 +1332,16 @@ internal HttpListenerContext HandleAuthentication(RequestContextBase memoryBlob,
                 }
             }
         }
+        
+        private static void FreeContext(ref HttpListenerContext httpContext, RequestContextBase memoryBlob)
+        {
+            if (httpContext != null)
+            {
+                httpContext.Request.DetachBlob(memoryBlob);
+                httpContext.Close();
+                httpContext = null;
+            }
+        }
```

Comment:
Don't know if you are worried about thread safety; but if you were you could do:
```csharp
private static void FreeContext(ref HttpListenerContext httpContext, RequestContextBase memoryBlob)
{
    var context = Interlocked.Exchange(ref httpContext, null);
    if (context != null)
    {
        context.Request.DetachBlob(memoryBlob);
        context.Close();
    }
}
```

Reply:
> Don't know if you are worried about thread safety

Is there reason to be?

Reply:
Don't know what bug 200343 was

Reply:
The purpose for `FreeContext()` is that, even if multiple threads call this function to free a HttpListenerContext, we want to make sure that particular HttpListenerContext is freed only once. Since we pass in the reference of the HttpListenerContext, and reads/writes to reference type is atomic, the null check at the beginning is sufficient. `Interlocked.Exchange` could impact performance to some extent.

> Don't know what bug 200343 was
.NET Framework VSO bug 200343.

Reply:
@benaadams 

>Don't know what bug 200343 was

This was an internal bug reported against .NET Framework.  It is fixed and will be available in .NET 4.7.2 release later this year.

Reply:
> Since we pass in the reference of the HttpListenerContext, and reads/writes to reference type is atomic, the null check at the beginning is sufficient. 

No, if that's the goal here, the interlocked is required.

Reply:
Ok.

Reply:
I looked at the code, I don't think the `HttpListenerContext` is shared between threads before `GetContext` returns (it is the only one calling `HandleAuthentication`).

Reply:
If it's not shared, then the interlocked isn't necessary, but then I'm also unclear what the bug is that this is fixing.

Reply:
> I don't think the HttpListenerContext is shared between threads before GetContext returns

It's shared. `HandleAuthentication` will also be called when [ListenerAsyncResult completed IO](https://github.com/dotnet/corefx/blob/1cce3631a78c11271f34762253b3a3273a8207fc/src/System.Net.HttpListener/src/System/Net/Windows/ListenerAsyncResult.Windows.cs#L47). So there is a race condition.

Reply:
I just glanced at the code. Isn't the variable being set to null a local? Is it closed over by something? Even if it's the same object, if the race is between two distinct references to the same object, nulling one out (regardless of interlocked) won't help.

Reply:
I missed the other call to `HandleAuthentication`.
However, the `HttpListenerContext` is created in that method and until we return the instance from `HandleAuthentication`, we do not share it with other threads. Isn't it that case?